File: java/data-structure/Stack/src/main/java/ru/reimu/stack/LinkedListStack.java
package ru.reimu.stack;

import ru.reimu.linkedList.LinkedList;

/**
 * @Author: Tomonori
 * @Date: 2020/2/20 14:37
 * @Title:
 * @Desc: ↓ ↓ ↓ ↓ ↓
 * -----
 */
public class LinkedListStack<T> implements Stack<T> {

    private LinkedList<T> linkedList;

    public LinkedListStack() {
        this.linkedList = new LinkedList<>();
    }

    @Override
    public int getSize() {
        return linkedList.getSize();
    }

    @Override
    public boolean isEmpty() {
        return linkedList.isEmpty();
    }

    @Override
    public void push(T t) {
        linkedList.addFirst(t);
    }

    @Override
    public T pop() {
        return linkedList.removeFirst();
    }

    @Override
    public T peek() {
        return linkedList.getFirst();
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        sb.append("Stack: top ");
        sb.append(linkedList);

        return sb.toString();
    }
}

File: golang/data-structure/stack/LinkedListStack.go
package stack

import (
	linkedList2 "all/linkedList"
	"bytes"
)

/**
 * @Author: Tomonori
 * @Date: 2020/2/20 14:56
 * @Title:
 * --- --- ---
 * @Desc:
 */

type LinkedListStack struct {
	linkedList *linkedList2.LinkedList
}

func NewLinkedListStack() *LinkedListStack {
	return &LinkedListStack{
		linkedList: linkedList2.NewLinkedList(),
	}
}

func (l *LinkedListStack) GetSize() int {
	return l.linkedList.GetSize()
}

func (l *LinkedListStack) IsEmpty() bool {
	return l.linkedList.IsEmpty()
}

func (l *LinkedListStack) Push(t interface{}) error {
	return l.linkedList.AddFirst(t)
}

func (l *LinkedListStack) Pop() (interface{}, error) {
	return l.linkedList.RemoveFirst()
}

func (l *LinkedListStack) Peek() (interface{}, error) {
	return l.linkedList.GetFirst()
}

func (l *LinkedListStack) String() string {
	buffer := bytes.Buffer{}
	buffer.WriteString("stack: top ")
	buffer.WriteString(l.linkedList.String())

	return buffer.String()
}
